File: app-tracker-automation/components/dropdown.py
"""
Dropdown component for handling select/dropdown elements.
Updated with CSS selectors from App Tracker HTML analysis.
Enhanced with robust synchronization for UAT environment.
"""

from playwright.sync_api import Page
from components.base_component import BaseComponent
from utils.logger import Logger


class Dropdown(BaseComponent):
    """Dropdown component class with App Tracker specific selectors"""
    
    def __init__(self, page: Page, locator: str):
        super().__init__(page, locator)
        self.logger = Logger()
    
    def _wait_for_loading_overlay_to_disappear(self, timeout=5000):
        """Wait for loading spinners, progress bars, or blocking overlays to disappear"""
        try:
            loading_selectors = [
                ".loading-overlay",
                ".spinner",
                ".progress-bar",
                "[class*='loading']",
                "[class*='spinner']",
                "[class*='overlay']",
                ".MuiCircularProgress-root",
                ".MuiBackdrop-root"
            ]
            
            for selector in loading_selectors:
                try:
                    loading_element = self.page.locator(selector).first
                    if loading_element.is_visible(timeout=1000):
                        self.logger.info(f"Waiting for loading element to disappear: {selector}")
                        loading_element.wait_for(state="hidden", timeout=timeout)
                        self.logger.info(f"Loading element disappeared: {selector}")
                except:
                    continue
        except Exception as e:
            self.logger.warning(f"Error waiting for loading overlay: {e}")
    
    def select_option(self, value: str):
        """Select option by value"""
        # Wait for network to settle before interaction
        self.page.wait_for_load_state("networkidle", timeout=10000)
        
        # Wait for element to be fully actionable
        self.element.wait_for(state="attached", timeout=5000)
        self.element.wait_for(state="visible", timeout=5000)
        self.element.wait_for(state="enabled", timeout=5000)
        
        # Wait for loading overlays to disappear
        self._wait_for_loading_overlay_to_disappear()
        
        self.element.select_option(value)
        
        # Wait for network to settle after selection
        self.page.wait_for_load_state("networkidle", timeout=5000)
    
    def select_option_by_label(self, label: str):
        """Select option by visible text"""
        # Wait for network to settle before interaction
        self.page.wait_for_load_state("networkidle", timeout=10000)
        
        # Wait for element to be fully actionable
        self.element.wait_for(state="attached", timeout=5000)
        self.element.wait_for(state="visible", timeout=5000)
        self.element.wait_for(state="enabled", timeout=5000)
        
        # Wait for loading overlays to disappear
        self._wait_for_loading_overlay_to_disappear()
        
        self.element.select_option(label=label)
        
        # Wait for network to settle after selection
        self.page.wait_for_load_state("networkidle", timeout=5000)
    
    def select_option_by_index(self, index: int):
        """Select option by index"""
        # Wait for network to settle before interaction
        self.page.wait_for_load_state("networkidle", timeout=10000)
        
        # Wait for element to be fully actionable
        self.element.wait_for(state="attached", timeout=5000)
        self.element.wait_for(state="visible", timeout=5000)
        self.element.wait_for(state="enabled", timeout=5000)
        
        # Wait for loading overlays to disappear
        self._wait_for_loading_overlay_to_disappear()
        
        self.element.select_option(index=index)
        
        # Wait for network to settle after selection
        self.page.wait_for_load_state("networkidle", timeout=5000)
    
    def get_selected_value(self) -> str:
        """Get selected option value"""
        return self.element.input_value()
    
    def get_selected_text(self) -> str:
        """Get selected option text"""
        return self.element.text_content()
    
    def get_all_options(self) -> list:
        """Get all available options"""
        return self.element.locator("option").all_inner_texts()
    
    def is_multiple(self) -> bool:
        """Check if dropdown allows multiple selection"""
        return self.element.get_attribute("multiple") is not None
    
    def clear_selection(self):
        """Clear selected options"""
        # Wait for network to settle before interaction
        self.page.wait_for_load_state("networkidle", timeout=10000)
        
        # Wait for element to be fully actionable
        self.element.wait_for(state="attached", timeout=5000)
        self.element.wait_for(state="visible", timeout=5000)
        self.element.wait_for(state="enabled", timeout=5000)
        
        # Wait for loading overlays to disappear
        self._wait_for_loading_overlay_to_disappear()
        
        self.element.select_option([])
        
        # Wait for network to settle after clearing
        self.page.wait_for_load_state("networkidle", timeout=5000)


# App Tracker Specific Dropdown Selectors
class AppTrackerDropdown(Dropdown):
    """App Tracker specific dropdown with Material-UI selectors"""
    
    # Filter Button Selector (from HTML)
    FILTER_BUTTON = "button.filterButton.MuiButton-textSizeSmall"
    
    # Sort Dropdown Selector (from HTML)
    SORT_DROPDOWN = "#mui-component-select-sortList"
    
    # Sort Options Container
    SORT_OPTIONS = "ul.MuiList-root[role='listbox']"
    
    def __init__(self, page: Page):
        super().__init__(page, self.FILTER_BUTTON)
    
    def click_filter_button(self):
        """Click the filter button to open filter dropdown"""
        # Wait for network to settle before interaction
        self.page.wait_for_load_state("networkidle", timeout=10000)
        
        filter_btn = self.page.locator(self.FILTER_BUTTON).first
        
        # Wait for element to be fully actionable
        filter_btn.wait_for(state="attached", timeout=5000)
        filter_btn.wait_for(state="visible", timeout=5000)
        filter_btn.wait_for(state="enabled", timeout=5000)
        
        # Wait for loading overlays to disappear
        self._wait_for_loading_overlay_to_disappear()
        
        filter_btn.click()
        
        # Wait for network to settle after click
        self.page.wait_for_load_state("networkidle", timeout=5000)
    
    def click_sort_dropdown(self):
        """Click the sort dropdown to open options"""
        # Wait for network to settle before interaction
        self.page.wait_for_load_state("networkidle", timeout=10000)
        
        sort_dropdown = self.page.locator(self.SORT_DROPDOWN).first
        
        # Wait for element to be fully actionable
        sort_dropdown.wait_for(state="attached", timeout=5000)
        sort_dropdown.wait_for(state="visible", timeout=5000)
        sort_dropdown.wait_for(state="enabled", timeout=5000)
        
        # Wait for loading overlays to disappear
        self._wait_for_loading_overlay_to_disappear()
        
        sort_dropdown.click()
        
        # Wait for network to settle after click
        self.page.wait_for_load_state("networkidle", timeout=5000)
    
    def select_sort_option(self, option_text: str):
        """Select a sort option by text"""
        # Wait for network to settle before interaction
        self.page.wait_for_load_state("networkidle", timeout=10000)
        
        option_locator = f"{self.SORT_OPTIONS} li:has-text('{option_text}')"
        option = self.page.locator(option_locator).first
        
        # Wait for element to be fully actionable
        option.wait_for(state="attached", timeout=5000)
        option.wait_for(state="visible", timeout=5000)
        option.wait_for(state="enabled", timeout=5000)
        
        # Wait for loading overlays to disappear
        self._wait_for_loading_overlay_to_disappear()
        
        option.click()
        
        # Wait for network to settle after selection
        self.page.wait_for_load_state("networkidle", timeout=5000)

File: ts-framework/components/dropdown.ts
import { Page, Locator, expect } from '@playwright/test';
import { BaseComponent } from './base_component';
import { Logger } from './logger';

/**
 * Dropdown component for handling select/dropdown elements.
 * Updated with CSS selectors from App Tracker HTML analysis.
 */
export class Dropdown extends BaseComponent {
    protected logger: Logger;

    constructor(page: Page, locator: string) {
        super(page, locator);
        this.logger = new Logger();
    }

    private async _waitForLoadingOverlayToDisappear(timeout: number = 5000): Promise<void> {
        try {
            const loadingSelectors = [
                ".loading-overlay",
                ".spinner",
                ".progress-bar",
                "[class*='loading']",
                "[class*='spinner']",
                "[class*='overlay']",
                ".MuiCircularProgress-root",
                ".MuiBackdrop-root"
            ];

            for (const selector of loadingSelectors) {
                try {
                    const loadingElement = this.page.locator(selector).first();
                    if (await loadingElement.isVisible({ timeout: 1000 })) {
                        this.logger.info(`Waiting for loading element to disappear: ${selector}`);
                        await loadingElement.waitFor({ state: "hidden", timeout });
                        this.logger.info(`Loading element disappeared: ${selector}`);
                    }
                } catch (e) {
                    continue;
                }
            }
        } catch (e) {
            this.logger.warning(`Error waiting for loading overlay: ${e}`);
        }
    }

    async selectOption(value: string): Promise<void> {
        await this.page.waitForLoadState("networkidle", { timeout: 10000 });
        await this.element.waitFor({ state: "attached", timeout: 5000 });
        await this.element.waitFor({ state: "visible", timeout: 5000 });
        
        await this._waitForLoadingOverlayToDisappear();
        await this.element.selectOption(value);
        await this.page.waitForLoadState("networkidle", { timeout: 5000 });
    }

    async selectOptionByLabel(label: string): Promise<void> {
        await this.page.waitForLoadState("networkidle", { timeout: 10000 });
        await this.element.waitFor({ state: "attached", timeout: 5000 });
        await this.element.waitFor({ state: "visible", timeout: 5000 });

        await this._waitForLoadingOverlayToDisappear();
        await this.element.selectOption({ label });
        await this.page.waitForLoadState("networkidle", { timeout: 5000 });
    }

    async selectOptionByIndex(index: number): Promise<void> {
        await this.page.waitForLoadState("networkidle", { timeout: 10000 });
        await this.element.waitFor({ state: "attached", timeout: 5000 });
        await this.element.waitFor({ state: "visible", timeout: 5000 });

        await this._waitForLoadingOverlayToDisappear();
        await this.element.selectOption({ index });
        await this.page.waitForLoadState("networkidle", { timeout: 5000 });
    }

    async getSelectedValue(): Promise<string> {
        return await this.element.inputValue();
    }

    async getSelectedText(): Promise<string | null> {
        return await this.element.textContent();
    }

    async getAllOptions(): Promise<string[]> {
        return await this.element.locator("option").allInnerTexts();
    }

    async isMultiple(): Promise<boolean> {
        return (await this.element.getAttribute("multiple")) !== null;
    }

    async clearSelection(): Promise<void> {
        await this.page.waitForLoadState("networkidle", { timeout: 10000 });
        await this.element.waitFor({ state: "attached", timeout: 5000 });
        await this.element.waitFor({ state: "visible", timeout: 5000 });

        await this._waitForLoadingOverlayToDisappear();
        await this.element.selectOption([]);
        await this.page.waitForLoadState("networkidle", { timeout: 5000 });
    }
}

/**
 * App Tracker Specific Dropdown Selectors
 */
export class AppTrackerDropdown extends Dropdown {
    static readonly FILTER_BUTTON = "button.filterButton.MuiButton-textSizeSmall";
    static readonly SORT_DROPDOWN = "#mui-component-select-sortList";
    static readonly SORT_OPTIONS = "ul.MuiList-root[role='listbox']";

    constructor(page: Page) {
        super(page, AppTrackerDropdown.FILTER_BUTTON);
    }

    async clickFilterButton(): Promise<void> {
        await this.page.waitForLoadState("networkidle", { timeout: 10000 });
        const filterBtn = this.page.locator(AppTrackerDropdown.FILTER_BUTTON).first();
        await filterBtn.waitFor({ state: "attached", timeout: 5000 });
        await filterBtn.waitFor({ state: "visible", timeout: 5000 });

        // Using private method from parent via type cast or changing access modifier if needed
        // Since it's private in Python, I made it private here too, but I'll change it to protected for subclasses.
        await (this as any)._waitForLoadingOverlayToDisappear();
        await filterBtn.click();
        await this.page.waitForLoadState("networkidle", { timeout: 5000 });
    }

    async clickSortDropdown(): Promise<void> {
        await this.page.waitForLoadState("networkidle", { timeout: 10000 });
        const sortDropdown = this.page.locator(AppTrackerDropdown.SORT_DROPDOWN).first();
        await sortDropdown.waitFor({ state: "attached", timeout: 5000 });
        await sortDropdown.waitFor({ state: "visible", timeout: 5000 });

        await (this as any)._waitForLoadingOverlayToDisappear();
        await sortDropdown.click();
        await this.page.waitForLoadState("networkidle", { timeout: 5000 });
    }

    async selectSortOption(optionText: string): Promise<void> {
        await this.page.waitForLoadState("networkidle", { timeout: 10000 });
        const optionLocator = `${AppTrackerDropdown.SORT_OPTIONS} li:has-text('${optionText}')`;
        const option = this.page.locator(optionLocator).first();
        
        await option.waitFor({ state: "attached", timeout: 5000 });
        await option.waitFor({ state: "visible", timeout: 5000 });

        await (this as any)._waitForLoadingOverlayToDisappear();
        await option.click();
        await this.page.waitForLoadState("networkidle", { timeout: 5000 });
    }
}
